### PR TITLE
Create a FileChooser component using showOpenFilePicker

### PR DESCRIPTION
**Pull request by iflan**
Before, `FileModal` always used a file input control for selecting
local files. This allowed the emulator to read from the file, but
precluded writing back to the file.

With this change, the `FileModal` delegates to the new `FileChooser`
component. The `FileChooser` will use `showOpenFilePicker` if it is
available and a regular file input if it's not.

Using `showOpenFilePicker` has the advantage of allowing the emulator
to write back to the file (if the user grants permission). While the
emulator does not yet take advantage of this write capability, that
will come.

### CODE REVIEW

**Comment by iflan** [comment]
I would just like to say that this is my first time trying to use React/Preact, so I may be doing everything wrong. Feedback is welcome!

File: js/components/FileChooser.tsx [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+            return;
+        }
+
+        if (fileHandlesRef.current?.length) {
+            filenameSpanRef.current.textContent = fileHandlesRef.current[0].name;
+        } else {
+            filenameSpanRef.current.textContent = 'No file selected';
+        }
+    }, [fileHandlesRef.current]);
+
```
Comment (whscullin) [inline js/components/FileChooser.tsx:147]: I would avoid using a ref to manipulate the DOM here, you can do something like:

```tsx
     const [selectedFilename, setSelectedFilename] = useState<string>()
```
```tsx
    setSelectedFilename(
        fileHandles.current?.length ?
        fileHandles.current?[0].name :
        'No file selected';
    )
```
```tsx
         <span role="label">{selectedFilename}</span>
```

File: test/env/jsdom-with-backdoors.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+                    [jsdomUtils.implSymbol]: impl,
+                });
+                impl[jsdomUtils.wrapperSymbol] = fileList;
+                const fileListCtor = this.global[jsdomUtils.ctorRegistrySymbol].FileList;
+                Object.setPrototypeOf(fileList, fileListCtor.prototype);
+                return fileList;
+            },
+        };
+    }
+}
```
Comment (whscullin) [inline test/env/jsdom-with-backdoors.js:34]: This looks like something I've had to do to deal with Location objects at one point.

**Review by whscullin — COMMENTED** [review]
It seems like you've got a pretty good handle on things, hooks can be a bit quirky and I think it might be possible to replace some of the `useRef()`s with `useState()`s but otherwise I think it looks good.

File: .eslintrc.json [hunk 3]
```diff
@@ -113,7 +113,8 @@
                 "node": true
             },
             "rules": {
-                "no-console": 0
+                "no-console": 0,
+                "@typescript-eslint/no-undef": "off",
```
Comment (whscullin) [inline .eslintrc.json:117]: Maybe this should be a more targeted exception, either just a comment in the file in question or declaring `backdoors` as a global for tests.

File: test/env/jsdom-with-backdoors.js (hunk 2 above)
Comment (iflan) [reply test/env/jsdom-with-backdoors.js:34]: This was super frustrating. It looks like jsdom is trying to be perfectly spec-compliant, but gives no way to set things up for tests. :-(

File: .eslintrc.json (hunk 3 above)
Comment (iflan) [reply .eslintrc.json]: I looked around a bunch for how to fix this "correctly" and this is the [advice I found](https://typescript-eslint.io/docs/linting/troubleshooting/#i-get-errors-from-the-no-undef-rule-about-global-variables-not-being-defined-even-though-there-are-no-typescript-errors):

> We strongly recommend that you do not use the `no-undef` lint rule on TypeScript projects. The checks it provides are already provided by TypeScript without the need for configuration - TypeScript just does this significantly better.

This is set up automatically if we use the configuration suggested by https://typescript-eslint.io/docs/linting/#configuration. Happy to do this in another change.

**Review by whscullin — APPROVED** [review]
This is cool, I've been meaning to look into the local filesystem api.

**Pull request merged**